File: src/types/documentoscopia.ts
/**
 * Espelho TS dos modelos do módulo Documentoscopia (Rust:
 * `models/documentoscopia.rs`). Coordenadas de bbox são NORMALIZADAS (0..1)
 * em relação à página. O arquivo original nunca é alterado — `relative_path`
 * aponta para a cópia preservada no workspace, com `sha256` próprio.
 */

export interface DocumentCaseFile {
  id: string;
  occurrence_id: string;
  title: string;
  original_filename: string;
  relative_path: string;
  /** "pdf" | "image" */
  file_type: string;
  extension: string;
  doc_type: string;
  sha256: string;
  size_bytes: number;
  page_count: number;
  has_text_layer: boolean;
  /** importado | ocr_pendente | ocr_concluido | revisado */
  status: string;
  metadata_json: string;
  notes: string;
  imported_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentPage {
  id: string;
  document_id: string;
  page_number: number;
  width: number;
  height: number;
  rotation: number;
  dpi: number | null;
  rendered_path: string | null;
  thumbnail_path: string | null;
  created_at: string;
}

export interface OcrRun {
  id: string;
  document_id: string;
  page_number: number | null;
  engine: string;
  engine_version: string;
  language: string;
  mode: string;
  status: string;
  avg_confidence: number | null;
  block_count: number;
  parameters_json: string;
  created_at: string;
}

export interface OcrTextBlock {
  id: string;
  ocr_run_id: string;
  document_id: string;
  page_number: number;
  text: string;
  confidence: number | null;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
  block_type: string;
  reading_order: number;
  corrected_text: string | null;
  reviewed: boolean;
  created_at: string;
}

export interface DetectedField {
  id: string;
  document_id: string;
  page_number: number | null;
  field_name: string;
  field_value: string;
  confidence: number | null;
  source: string;
  bbox_x: number | null;
  bbox_y: number | null;
  bbox_w: number | null;
  bbox_h: number | null;
  reviewed: boolean;
  corrected_value: string | null;
  created_at: string;
}

export interface DocumentRegion {
  id: string;
  document_id: string;
  page_number: number;
  region_type: string;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
  label: string;
  confidence: number | null;
  notes: string;
  created_at: string;
}

export interface DocumentAnalysis {
  id: string;
  document_id: string;
  analysis_type: string;
  status: string;
  parameters_json: string;
  result_json: string;
  summary: string;
  created_at: string;
}

export interface ComparisonSession {
  id: string;
  occurrence_id: string;
  questioned_document_id: string;
  reference_document_id: string;
  comparison_type: string;
  results_json: string;
  summary: string;
  created_at: string;
}

export interface DocumentLog {
  id: string;
  document_id: string | null;
  occurrence_id: string;
  action: string;
  parameters_json: string;
  result: string;
  source_hash: string | null;
  output_hash: string | null;
  actor: string | null;
  created_at: string;
}

export interface OcrRunResult {
  run: OcrRun;
  blocks: OcrTextBlock[];
}

// --- Inputs (snake_case — batem com os structs serde do backend) ---

export interface TextBlockInput {
  page_number: number;
  text: string;
  confidence: number | null;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
  block_type: string;
  reading_order: number;
}

export interface OcrRunInput {
  page_number: number | null;
  engine: string;
  engine_version: string;
  language: string;
  mode: string;
  parameters_json: string;
  blocks: TextBlockInput[];
}

export interface FieldInput {
  page_number: number | null;
  field_name: string;
  field_value: string;
  confidence: number | null;
  source: string;
  bbox_x: number | null;
  bbox_y: number | null;
  bbox_w: number | null;
  bbox_h: number | null;
}

export interface RegionInput {
  page_number: number;
  region_type: string;
  bbox_x: number;
  bbox_y: number;
  bbox_w: number;
  bbox_h: number;
  label: string;
  confidence: number | null;
  notes: string;
}

/** Tipos de documento oferecidos na UI (rótulo + valor persistido). */
export const DOC_TYPES: { value: string; label: string }[] = [
  { value: "cnh", label: "CNH" },
  { value: "rg", label: "RG" },
  { value: "crlv", label: "CRLV" },
  { value: "contrato", label: "Contrato" },
  { value: "recibo", label: "Recibo" },
  { value: "declaracao", label: "Declaração" },
  { value: "oficio", label: "Ofício" },
  { value: "boletim", label: "Boletim" },
  { value: "laudo", label: "Laudo externo" },
  { value: "processo", label: "Processo" },
  { value: "outro", label: "Outro" },
];

export function docTypeLabel(value: string): string {
  return DOC_TYPES.find((t) => t.value === value)?.label ?? "Outro";
}

/** Rótulo + cor (token CSS) de um status de documento. */
export function docStatusInfo(status: string): { label: string; tone: string } {
  switch (status) {
    case "importado":
      return { label: "Importado", tone: "info" };
    case "ocr_pendente":
      return { label: "OCR pendente", tone: "warn" };
    case "ocr_concluido":
      return { label: "OCR concluído", tone: "accent" };
    case "revisado":
      return { label: "Revisado", tone: "success" };
    default:
      return { label: status, tone: "muted" };
  }
}

File: src-tauri/src/models/documentoscopia.rs
//! Modelos do módulo Documentoscopia.
//!
//! Espelham as tabelas de `migrations/016_documentoscopia.sql`. Coordenadas de
//! bounding box são NORMALIZADAS (0..1) em relação à página. O arquivo original
//! nunca é alterado — `relative_path` aponta para a cópia preservada no
//! workspace, com `sha256` próprio (cadeia de custódia).

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Um documento (PDF ou imagem) importado para análise documentoscópica.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentCaseFile {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub title: String,
    pub original_filename: String,
    pub relative_path: String,
    /// "pdf" | "image".
    pub file_type: String,
    pub extension: String,
    /// cnh|rg|crlv|contrato|recibo|declaracao|oficio|boletim|laudo|processo|outro
    pub doc_type: String,
    pub sha256: String,
    pub size_bytes: u64,
    pub page_count: i64,
    pub has_text_layer: bool,
    /// importado|ocr_pendente|ocr_concluido|revisado
    pub status: String,
    pub metadata_json: String,
    pub notes: String,
    pub imported_by: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Uma página renderizada (imagem = 1 página; PDF = N páginas).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentPage {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_number: i64,
    pub width: i64,
    pub height: i64,
    pub rotation: i64,
    pub dpi: Option<i64>,
    pub rendered_path: Option<String>,
    pub thumbnail_path: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// Uma execução de OCR / extração de texto (qualquer motor, inclusive a camada
/// textual de PDF digital ou o motor mock).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OcrRun {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_number: Option<i64>,
    pub engine: String,
    pub engine_version: String,
    pub language: String,
    pub mode: String,
    pub status: String,
    pub avg_confidence: Option<f64>,
    pub block_count: i64,
    pub parameters_json: String,
    pub created_at: DateTime<Utc>,
}

/// Um bloco de texto extraído (com bbox normalizado e confiança).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OcrTextBlock {
    pub id: Uuid,
    pub ocr_run_id: Uuid,
    pub document_id: Uuid,
    pub page_number: i64,
    pub text: String,
    pub confidence: Option<f64>,
    pub bbox_x: f64,
    pub bbox_y: f64,
    pub bbox_w: f64,
    pub bbox_h: f64,
    pub block_type: String,
    pub reading_order: i64,
    pub corrected_text: Option<String>,
    pub reviewed: bool,
    pub created_at: DateTime<Utc>,
}

/// Um campo extraído por heurística/OCR/manual, com revisão humana.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectedField {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_number: Option<i64>,
    pub field_name: String,
    pub field_value: String,
    pub confidence: Option<f64>,
    pub source: String,
    pub bbox_x: Option<f64>,
    pub bbox_y: Option<f64>,
    pub bbox_w: Option<f64>,
    pub bbox_h: Option<f64>,
    pub reviewed: bool,
    pub corrected_value: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// Uma região marcada sobre o documento (detectada ou manual).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentRegion {
    pub id: Uuid,
    pub document_id: Uuid,
    pub page_number: i64,
    pub region_type: String,
    pub bbox_x: f64,
    pub bbox_y: f64,
    pub bbox_w: f64,
    pub bbox_h: f64,
    pub label: String,
    pub confidence: Option<f64>,
    pub notes: String,
    pub created_at: DateTime<Utc>,
}

/// Uma análise documentoscópica assistida (ELA, ruído, integridade…).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentAnalysis {
    pub id: Uuid,
    pub document_id: Uuid,
    pub analysis_type: String,
    pub status: String,
    pub parameters_json: String,
    pub result_json: String,
    pub summary: String,
    pub created_at: DateTime<Utc>,
}

/// Uma sessão de comparação entre documento questionado e padrão.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComparisonSession {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub questioned_document_id: Uuid,
    pub reference_document_id: Uuid,
    pub comparison_type: String,
    pub results_json: String,
    pub summary: String,
    pub created_at: DateTime<Utc>,
}

/// Uma entrada do histórico/auditoria do documento.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentLog {
    pub id: Uuid,
    pub document_id: Option<Uuid>,
    pub occurrence_id: Uuid,
    pub action: String,
    pub parameters_json: String,
    pub result: String,
    pub source_hash: Option<String>,
    pub output_hash: Option<String>,
    pub actor: Option<String>,
    pub created_at: DateTime<Utc>,
}
